apps/website/src/components/events/virtual-ticket/TicketEditor.tsx

```diff
@@ -0,0 +1,354 @@
+    saveStickers: () => {
+      dispatch({ type: "save" });
+    },
+    resetStickers,
+  };
+}
+
+export function useTicketEditorContext() {
+  return useContext(ticketEditorContext);
+}
+
+export function TicketEditor({
+  children,
+  width = defaultTicketWidth,
+  height = defaultTicketHeight,
+  canvasWidth = width,
+  canvasHeight = height,
+  onSave,
+  stickerPack,
+  stickerData,
+}: {
+  children?: ReactNode;
+  width?: number;
+  height?: number;
+  canvasWidth?: number;
+  canvasHeight?: number;
+  stickerData: ReturnType<typeof useStickerData>;
+  onSave: () => void;
+  stickerPack: StickerPack;
+}) {
+  const { removeSticker, addSticker, stickers, isDirty, resetStickers } =
+    stickerData;
+  const [selectedSticker, setSelectedSticker] = useState<string | null>(null);
+  const contextValue = useMemo(
+    () => ({
+      canvasWidth,
+      canvasHeight,
+      selectedSticker,
+      setSelectedSticker,
+    }),
+    [canvasHeight, canvasWidth, selectedSticker],
+  );
+
+  return (
+    <ticketEditorContext.Provider value={contextValue}>
+      <div className="-mx-4 w-[calc(100%+2rem)] overflow-x-auto outline-1 outline-red md:w-full">
+        <div
+          className="mx-auto flex w-[calc(900px+3rem)] flex-col items-center px-6 py-12 drop-shadow-xl"
+          onClick={() => setSelectedSticker(null)}
+        >
+          {children}
+        </div>
+      </div>
+
+      <div className="relative z-40 -mt-14 flex gap-2 rounded-2xl bg-white p-1 shadow-xl">
+        <Button
+          className={`whitespace-nowrap ${
+            isDirty ? "bg-green-600 text-white" : "bg-gray-300 text-gray-900"
+          }`}
+          width="auto"
+          size="small"
+          disabled={!isDirty}
+          onClick={() => {
+            onSave();
+          }}
+        >
+          {isDirty ? (
+            "Save ticket"
+          ) : (
+            <>
+              <IconCheck className="h-6 w-6" />
+              Saved
+            </>
+          )}
+        </Button>
+
+        <Button
+          className="whitespace-nowrap"
+          width="auto"
+          size="small"
+          disabled={!selectedSticker}
+          onClick={() => {
+            if (selectedSticker) {
+              removeSticker(selectedSticker);
+              setSelectedSticker(null);
+            }
+          }}
+        >
+          <IconMinusCircle className="h-6 w-6" />
+          Remove sticker
+        </Button>
+
+        <Button
+          className="whitespace-nowrap"
+          width="auto"
+          size="small"
+          confirmationMessage="Are you sure you want to remove all stickers?"
+          onClick={resetStickers}
+        >
+          <IconTrash className="h-6 w-6" />
+          Reset
+        </Button>
+      </div>
+
+      <div className="mx-auto mt-8 flex max-w-3xl flex-row gap-4">
+        {typeSafeObjectKeys(stickerPack.groups).map((groupId) => {
+          const group = stickerPack.groups[groupId]!;
+
+          return (
```

Comment:
maybe refactor a bit so its not nested that many levels. maybe extract a component

Reply:
Agree, would probably be nice to have a StickerPack component, could just be in this file